File: go/discovery.go
package kumiho

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const (
	defaultControlPlane = "https://control.kumiho.cloud"
	defaultCacheKey     = "__default__"
	defaultLocalCEPort  = 9190
)

// DiscoveryError indicates a control-plane discovery / bootstrap failure.
type DiscoveryError struct{ Msg string }

func (e *DiscoveryError) Error() string { return "discovery: " + e.Msg }

func (e *DiscoveryError) kumihoError() {}

// RegionRouting is the regional gRPC routing returned by the control plane.
type RegionRouting struct {
	RegionCode    string `json:"region_code"`
	ServerURL     string `json:"server_url"`
	GRPCAuthority string `json:"grpc_authority,omitempty"`
}

// CacheControl is the cache window emitted by the control plane.
type CacheControl struct {
	IssuedAt            string `json:"issued_at"`
	RefreshAt           string `json:"refresh_at"`
	ExpiresAt           string `json:"expires_at"`
	ExpiresInSeconds    int64  `json:"expires_in_seconds"`
	RefreshAfterSeconds int64  `json:"refresh_after_seconds"`
}

func parseTS(raw string) (time.Time, bool) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return time.Time{}, false
	}
	if strings.HasSuffix(raw, "Z") {
		raw = strings.TrimSuffix(raw, "Z") + "+00:00"
	}
	if t, err := time.Parse(time.RFC3339, raw); err == nil {
		return t.UTC(), true
	}
	if t, err := time.Parse("2006-01-02T15:04:05", raw); err == nil {
		return t.UTC(), true
	}
	return time.Time{}, false
}

func (c CacheControl) isExpired() bool {
	t, ok := parseTS(c.ExpiresAt)
	return !ok || !time.Now().UTC().Before(t)
}

func (c CacheControl) shouldRefresh() bool {
	t, ok := parseTS(c.RefreshAt)
	return !ok || !time.Now().UTC().Before(t)
}

// DiscoveryRecord is a resolved tenant routing record.
type DiscoveryRecord struct {
	TenantID     string          `json:"tenant_id"`
	TenantName   string          `json:"tenant_name,omitempty"`
	Roles        []string        `json:"roles,omitempty"`
	Guardrails   json.RawMessage `json:"guardrails,omitempty"`
	Region       RegionRouting   `json:"region"`
	CacheControl CacheControl    `json:"cache_control"`
}

// Target is the gRPC endpoint to dial (authority preferred over server URL).
func (r *DiscoveryRecord) Target() string {
	if r.Region.GRPCAuthority != "" {
		return r.Region.GRPCAuthority
	}
	return r.Region.ServerURL
}

// ---- at-rest cache encryption (machine-bound, defense-in-depth) ----

func machineID() string {
	switch runtime.GOOS {
	case "linux":
		for _, p := range []string{"/etc/machine-id", "/var/lib/dbus/machine-id"} {
			if b, err := os.ReadFile(p); err == nil {
				if s := strings.TrimSpace(string(b)); s != "" {
					return s
				}
			}
		}
	case "darwin":
		if out, err := exec.Command("ioreg", "-rd1", "-c", "IOPlatformExpertDevice").Output(); err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				if strings.Contains(line, "IOPlatformUUID") {
					parts := strings.Split(line, `"`)
					if len(parts) >= 2 {
						return parts[len(parts)-2]
					}
				}
			}
		}
	}
	idFile := filepath.Join(configDir(), ".machine_id")
	if b, err := os.ReadFile(idFile); err == nil {
		if s := strings.TrimSpace(string(b)); s != "" {
			return s
		}
	}
	buf := make([]byte, 16)
	_, _ = rand.Read(buf)
	id := fmt.Sprintf("%x", buf)
	_ = os.MkdirAll(configDir(), 0o700)
	_ = os.WriteFile(idFile, []byte(id), 0o600)
	return id
}

func deriveKey() []byte {
	login := os.Getenv("USER")
	if login == "" {
		login = os.Getenv("LOGNAME")
	}
	uid := strconv.Itoa(os.Getuid())
	material := fmt.Sprintf("kumiho-discovery-cache-v1:%s:%s%s", machineID(), login, uid)
	sum := sha256.Sum256([]byte(material))
	return sum[:]
}

func keystream(key, iv []byte, n int) []byte {
	first := sha256.Sum256(append(append([]byte{}, key...), iv...))
	stream := append([]byte{}, first[:]...)
	for len(stream) < n {
		tail := stream[len(stream)-32:]
		next := sha256.Sum256(append(append([]byte{}, key...), tail...))
		stream = append(stream, next[:]...)
	}
	return stream
}

func encryptCache(plaintext string) string {
	key := deriveKey()
	iv := make([]byte, 16)
	_, _ = rand.Read(iv)
	pt := []byte(plaintext)
	ks := keystream(key, iv, len(pt))
	ct := make([]byte, len(pt))
	for i := range pt {
		ct[i] = pt[i] ^ ks[i]
	}
	mac := hmac.New(sha256.New, key)
	mac.Write(iv)
	mac.Write(ct)
	tag := mac.Sum(nil)[:16]
	blob := append(append(append([]byte{}, iv...), ct...), tag...)
	return "enc:v1:" + base64.StdEncoding.EncodeToString(blob)
}

func decryptCache(encrypted string) (string, bool) {
	rest, ok := strings.CutPrefix(encrypted, "enc:v1:")
	if !ok {
		return encrypted, true // legacy plaintext
	}
	key := deriveKey()
	raw, err := base64.StdEncoding.DecodeString(rest)
	if err != nil || len(raw) < 32 {
		return "", false
	}
	iv, ct, tag := raw[:16], raw[16:len(raw)-16], raw[len(raw)-16:]
	mac := hmac.New(sha256.New, key)
	mac.Write(iv)
	mac.Write(ct)
	if !hmac.Equal(mac.Sum(nil)[:16], tag) {
		return "", false
	}
	ks := keystream(key, iv, len(ct))
	pt := make([]byte, len(ct))
	for i := range ct {
		pt[i] = ct[i] ^ ks[i]
	}
	return string(pt), true
}

func cachePath() string {
	if p := os.Getenv("KUMIHO_DISCOVERY_CACHE_FILE"); p != "" {
		return p
	}
	// Namespaced to avoid clashing with other SDKs' caches on the same key.
	return filepath.Join(configDir(), "discovery-cache.go.json")
}

func cacheReadAll() map[string]DiscoveryRecord {
	out := map[string]DiscoveryRecord{}
	data, err := os.ReadFile(cachePath())
	if err != nil {
		return out
	}
	plain, ok := decryptCache(string(data))
	if !ok {
		return out
	}
	_ = json.Unmarshal([]byte(plain), &out)
	return out
}

func cacheStore(key string, rec DiscoveryRecord) {
	all := cacheReadAll()
	all[key] = rec
	data, err := json.MarshalIndent(all, "", "  ")
	if err != nil {
		return
	}
	_ = os.MkdirAll(filepath.Dir(cachePath()), 0o700)
	_ = os.WriteFile(cachePath(), []byte(encryptCache(string(data))), 0o600)
}

func buildDiscoveryURL(base string) string {
	base = strings.TrimRight(base, "/")
	switch {
	case strings.HasSuffix(base, "/api/discovery/tenant"):
		return base
	case strings.HasSuffix(base, "/api/discovery"):
		return base + "/tenant"
	case strings.HasSuffix(base, "/api"):
		return base + "/discovery/tenant"
	default:
		return base + "/api/discovery/tenant"
	}
}

func fetchRemote(ctx context.Context, baseURL, idToken, tenantHint string) (DiscoveryRecord, error) {
	body := map[string]string{}
	if tenantHint != "" {
		body["tenant_hint"] = tenantHint
	}
	payload, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, buildDiscoveryURL(baseURL), bytes.NewReader(payload))
	if err != nil {
		return DiscoveryRecord{}, &DiscoveryError{Msg: err.Error()}
	}
	req.Header.Set("Authorization", "Bearer "+idToken)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "kumiho-go/"+Version)
	resp, err := (&http.Client{Timeout: discoveryTimeout()}).Do(req)
	if err != nil {
		return DiscoveryRecord{}, &DiscoveryError{Msg: err.Error()}
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return DiscoveryRecord{}, &DiscoveryError{Msg: fmt.Sprintf("discovery endpoint returned %d", resp.StatusCode)}
	}
	var rec DiscoveryRecord
	if err := json.NewDecoder(resp.Body).Decode(&rec); err != nil {
		return DiscoveryRecord{}, &DiscoveryError{Msg: "invalid discovery payload: " + err.Error()}
	}
	return rec, nil
}

// fetchFresh fetches a discovery record, trying each token candidate in turn
// (the bearer token, plus a Firebase fallback when it's a control-plane token)
// and returning the last error if all fail. Mirrors Python's fetch_fresh.
func fetchFresh(ctx context.Context, base, idToken, tenantHint string) (DiscoveryRecord, error) {
	var lastErr error
	for _, tok := range discoveryTokenCandidates(idToken) {
		rec, err := fetchRemote(ctx, base, tok, tenantHint)
		if err == nil {
			return rec, nil
		}
		lastErr = err
	}
	if lastErr != nil {
		return DiscoveryRecord{}, lastErr
	}
	return DiscoveryRecord{}, &DiscoveryError{Msg: "discovery failed without a usable bearer token"}
}

// resolveDiscovery resolves a DiscoveryRecord, using the encrypted cache when fresh.
func resolveDiscovery(ctx context.Context, idToken, tenantHint string, forceRefresh bool) (DiscoveryRecord, error) {
	base := os.Getenv("KUMIHO_CONTROL_PLANE_URL")
	if base == "" {
		base = defaultControlPlane
	}
	key := tenantHint
	if key == "" {
		key = defaultCacheKey
	}

	if !forceRefresh {
		if cached, ok := cacheReadAll()[key]; ok && !cached.CacheControl.isExpired() {
			if cached.CacheControl.shouldRefresh() {
				if fresh, err := fetchFresh(ctx, base, idToken, tenantHint); err == nil {
					cacheStore(key, fresh)
					return fresh, nil
				} else if !cached.CacheControl.isExpired() {
					return cached, nil
				} else {
					return DiscoveryRecord{}, err
				}
			}
			return cached, nil
		}
	}
	fresh, err := fetchFresh(ctx, base, idToken, tenantHint)
	if err != nil {
		return DiscoveryRecord{}, err
	}
	cacheStore(key, fresh)
	return fresh, nil
}

// TenantInfo returns the cached discovery record for the given tenant hint (or
// the default tenant), or nil if no cache entry exists. Mirrors Python
// get_tenant_info (tenant id, name, roles, region, guardrails).
func TenantInfo(tenantHint string) *DiscoveryRecord {
	key := tenantHint
	if key == "" {
		key = defaultCacheKey
	}
	if rec, ok := cacheReadAll()[key]; ok {
		return &rec
	}
	return nil
}

// TenantSlug returns a URL-safe tenant slug (or a shortened tenant id) for the
// given tenant hint, or "" if no cached tenant info exists. Mirrors Python
// get_tenant_slug.
func TenantSlug(tenantHint string) string {
	rec := TenantInfo(tenantHint)
	if rec == nil {
		return ""
	}
	if rec.TenantName != "" && isURLSafeSlug(rec.TenantName) {
		return strings.ToLower(rec.TenantName)
	}
	if rec.TenantID != "" {
		return strings.SplitN(rec.TenantID, "-", 2)[0]
	}
	return rec.TenantName
}

func isURLSafeSlug(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if !((r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') || r == '-') {
			return false
		}
	}
	return true
}

// resolveLocalCEEndpoint probes loopback ports for a self-hosted CE server.
// It returns an error when KUMIHO_LOCAL_SERVER_ENDPOINT / KUMIHO_LOCAL_SERVER_PORT
// is set to a non-loopback or otherwise invalid value, mirroring the Python SDK.
func resolveLocalCEEndpoint(ctx context.Context) (string, error) {
	var candidates []string
	if ep := strings.TrimSpace(os.Getenv("KUMIHO_LOCAL_SERVER_ENDPOINT")); ep != "" {
		t, err := normalizeLocalTarget(ep)
		if err != nil {
			return "", err
		}
		candidates = []string{t}
	} else if port := strings.TrimSpace(os.Getenv("KUMIHO_LOCAL_SERVER_PORT")); port != "" {
		n, err := strconv.Atoi(port)
		if err != nil || n <= 0 || n > 65535 {
			return "", &DiscoveryError{Msg: "KUMIHO_LOCAL_SERVER_PORT must be a numeric loopback port"}
		}
		candidates = []string{fmt.Sprintf("127.0.0.1:%d", n)}
	} else {
		candidates = []string{fmt.Sprintf("127.0.0.1:%d", defaultLocalCEPort)}
	}
	for _, t := range candidates {
		if probeCE(ctx, t) {
			return t, nil
		}
	}
	return "", nil
}

// normalizeLocalTarget strips the scheme/path from a local CE endpoint and
// enforces that it points at a loopback host (localhost, 127.0.0.1, ::1) so a
// tokenless client can never be routed to a remote server. Mirrors the Python
// _normalise_local_ce_target guard.
func normalizeLocalTarget(raw string) (string, error) {
	raw = strings.TrimSpace(raw)
	if i := strings.Index(raw, "://"); i >= 0 {
		raw = raw[i+3:]
	}
	if i := strings.IndexByte(raw, '/'); i >= 0 {
		raw = raw[:i] // strip any trailing path
	}
	host := raw
	port := defaultLocalCEPort
	if h, p, err := net.SplitHostPort(raw); err == nil {
		host = h
		n, perr := strconv.Atoi(p)
		if perr != nil || n <= 0 || n > 65535 {
			return "", &DiscoveryError{Msg: "KUMIHO_LOCAL_SERVER_ENDPOINT port must be between 1 and 65535"}
		}
		port = n
	} else {
		host = strings.TrimSuffix(strings.TrimPrefix(host, "["), "]") // bare/bracketed IPv6
	}
	if host == "" {
		return "", &DiscoveryError{Msg: "KUMIHO_LOCAL_SERVER_ENDPOINT must include a loopback host"}
	}
	if !isLoopbackHost(host) {
		return "", &DiscoveryError{Msg: "KUMIHO_LOCAL_SERVER_ENDPOINT must point to localhost, 127.0.0.1, or ::1"}
	}
	return net.JoinHostPort(host, strconv.Itoa(port)), nil
}

func isLoopbackHost(host string) bool {
	if strings.EqualFold(host, "localhost") {
		return true
	}
	if ip := net.ParseIP(host); ip != nil {
		return ip.IsLoopback()
	}
	return false
}

// discoveryTimeout is the control-plane HTTP timeout (KUMIHO_DISCOVERY_TIMEOUT_SECONDS,
// default 10s).
func discoveryTimeout() time.Duration {
	if v := os.Getenv("KUMIHO_DISCOVERY_TIMEOUT_SECONDS"); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil && f > 0 {
			return time.Duration(f * float64(time.Second))
		}
	}
	return 10 * time.Second
}

// localCETimeout is the local-CE probe timeout (KUMIHO_LOCAL_DISCOVERY_TIMEOUT_SECONDS,
// default 0.5s, floor 0.05s).
func localCETimeout() time.Duration {
	if v := os.Getenv("KUMIHO_LOCAL_DISCOVERY_TIMEOUT_SECONDS"); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			if f < 0.05 {
				f = 0.05
			}
			return time.Duration(f * float64(time.Second))
		}
	}
	return 500 * time.Millisecond
}

func probeCE(ctx context.Context, target string) bool {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "http://"+target+"/api/_live", nil)
	if err != nil {
		return false
	}
	resp, err := (&http.Client{Timeout: localCETimeout()}).Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return false
	}
	var body map[string]any
	if json.NewDecoder(resp.Body).Decode(&body) != nil {
		return false
	}
	mode, _ := body["deployment_mode"].(string)
	return mode == "self_hosted_ce"
}

File: rust/src/discovery.rs
//! Control-plane discovery, encrypted routing cache, and local-CE fallback.
//!
//! Mirrors the Python `discovery.py`: resolve the regional gRPC endpoint for a
//! tenant via `POST {control_plane}/api/discovery/tenant`, cache the routing
//! payload at rest (XOR + HMAC, machine-derived key), and honour the
//! cache-control refresh/expiry window. Also probes a loopback self-hosted CE
//! server so a local dev deployment "just works" without a token.

use base64::Engine;
use chrono::{DateTime, Utc};
use hmac::{Hmac, Mac};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::path::PathBuf;

type HmacSha256 = Hmac<Sha256>;

const DEFAULT_CONTROL_PLANE: &str = "https://control.kumiho.cloud";
const DEFAULT_CACHE_KEY: &str = "__default__";
const DEFAULT_LOCAL_CE_PORT: u16 = 9190;

/// Discovery / control-plane bootstrap failure.
#[derive(Debug, Clone, thiserror::Error)]
#[error("{0}")]
pub struct DiscoveryError(pub String);

impl From<DiscoveryError> for crate::Error {
    fn from(e: DiscoveryError) -> Self {
        crate::Error::Discovery(e.0)
    }
}

/// Regional gRPC routing returned by the control plane.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegionRouting {
    pub region_code: String,
    pub server_url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub grpc_authority: Option<String>,
}

/// Cache-control window emitted by the control plane.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheControl {
    pub issued_at: String,
    pub refresh_at: String,
    pub expires_at: String,
    #[serde(default)]
    pub expires_in_seconds: i64,
    #[serde(default)]
    pub refresh_after_seconds: i64,
}

impl CacheControl {
    fn ts(raw: &str) -> Option<DateTime<Utc>> {
        DateTime::parse_from_rfc3339(&raw.replace(' ', "T"))
            .ok()
            .map(|d| d.with_timezone(&Utc))
    }
    fn is_expired(&self) -> bool {
        match Self::ts(&self.expires_at) {
            Some(t) => Utc::now() >= t,
            None => true,
        }
    }
    fn should_refresh(&self) -> bool {
        match Self::ts(&self.refresh_at) {
            Some(t) => Utc::now() >= t,
            None => true,
        }
    }
}

/// A resolved discovery record (tenant + routing + cache window).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryRecord {
    pub tenant_id: String,
    #[serde(default)]
    pub tenant_name: Option<String>,
    #[serde(default)]
    pub roles: Vec<String>,
    #[serde(default)]
    pub guardrails: Option<serde_json::Value>,
    pub region: RegionRouting,
    pub cache_control: CacheControl,
}

impl DiscoveryRecord {
    /// The gRPC target to dial (authority preferred over server URL).
    pub fn target(&self) -> String {
        self.region
            .grpc_authority
            .clone()
            .unwrap_or_else(|| self.region.server_url.clone())
    }
}

// ---- at-rest cache encryption (defense-in-depth, machine-bound) ----

fn machine_id() -> String {
    #[cfg(target_os = "linux")]
    {
        for path in ["/etc/machine-id", "/var/lib/dbus/machine-id"] {
            if let Ok(s) = std::fs::read_to_string(path) {
                let s = s.trim();
                if !s.is_empty() {
                    return s.to_string();
                }
            }
        }
    }
    #[cfg(target_os = "macos")]
    {
        if let Ok(out) = std::process::Command::new("ioreg")
            .args(["-rd1", "-c", "IOPlatformExpertDevice"])
            .output()
        {
            let text = String::from_utf8_lossy(&out.stdout);
            for line in text.lines() {
                if line.contains("IOPlatformUUID") {
                    if let Some(uuid) = line.split('"').nth_back(1) {
                        return uuid.to_string();
                    }
                }
            }
        }
    }
    // Fallback: a random id stored in the config dir.
    let id_file = crate::token_loader::config_dir().join(".machine_id");
    if let Ok(s) = std::fs::read_to_string(&id_file) {
        let s = s.trim();
        if !s.is_empty() {
            return s.to_string();
        }
    }
    let new_id = uuid::Uuid::new_v4().to_string();
    let _ = std::fs::create_dir_all(crate::token_loader::config_dir());
    let _ = std::fs::write(&id_file, &new_id);
    new_id
}

fn derive_key() -> [u8; 32] {
    let login = std::env::var("USER").or_else(|_| std::env::var("LOGNAME")).unwrap_or_default();
    let material = format!("kumiho-discovery-cache-v1:{}:{}", machine_id(), login);
    let mut hasher = Sha256::new();
    hasher.update(material.as_bytes());
    hasher.finalize().into()
}

fn keystream(key: &[u8], iv: &[u8], len: usize) -> Vec<u8> {
    let mut stream: Vec<u8> = {
        let mut h = Sha256::new();
        h.update(key);
        h.update(iv);
        h.finalize().to_vec()
    };
    while stream.len() < len {
        let tail = stream[stream.len() - 32..].to_vec();
        let mut h = Sha256::new();
        h.update(key);
        h.update(&tail);
        stream.extend_from_slice(&h.finalize());
    }
    stream
}

fn encrypt(plaintext: &str) -> String {
    let key = derive_key();
    let iv: [u8; 16] = rand::random();
    let pt = plaintext.as_bytes();
    let ks = keystream(&key, &iv, pt.len());
    let ct: Vec<u8> = pt.iter().zip(ks.iter()).map(|(p, k)| p ^ k).collect();
    let mut mac = HmacSha256::new_from_slice(&key).expect("hmac key");
    mac.update(&iv);
    mac.update(&ct);
    let tag = &mac.finalize().into_bytes()[..16];
    let mut blob = Vec::with_capacity(16 + ct.len() + 16);
    blob.extend_from_slice(&iv);
    blob.extend_from_slice(&ct);
    blob.extend_from_slice(tag);
    format!("enc:v1:{}", base64::engine::general_purpose::STANDARD.encode(blob))
}

fn decrypt(encrypted: &str) -> Option<String> {
    let Some(b64) = encrypted.strip_prefix("enc:v1:") else {
        return Some(encrypted.to_string()); // legacy plaintext
    };
    let key = derive_key();
    let raw = base64::engine::general_purpose::STANDARD.decode(b64).ok()?;
    if raw.len() < 32 {
        return None;
    }
    let iv = &raw[..16];
    let tag = &raw[raw.len() - 16..];
    let ct = &raw[16..raw.len() - 16];
    let mut mac = HmacSha256::new_from_slice(&key).ok()?;
    mac.update(iv);
    mac.update(ct);
    let expected = &mac.finalize().into_bytes()[..16];
    if expected != tag {
        return None;
    }
    let ks = keystream(&key, iv, ct.len());
    let pt: Vec<u8> = ct.iter().zip(ks.iter()).map(|(c, k)| c ^ k).collect();
    String::from_utf8(pt).ok()
}

fn cache_path() -> PathBuf {
    if let Ok(p) = std::env::var("KUMIHO_DISCOVERY_CACHE_FILE") {
        if !p.is_empty() {
            return PathBuf::from(p);
        }
    }
    // Own cache file (the Python SDK uses discovery-cache.json on the same key
    // derivation; we namespace ours to avoid cross-language key clashes).
    crate::token_loader::config_dir().join("discovery-cache.rust.json")
}

struct DiscoveryCache {
    path: PathBuf,
}

impl DiscoveryCache {
    fn read_all(&self) -> HashMap<String, DiscoveryRecord> {
        let Ok(content) = std::fs::read_to_string(&self.path) else {
            return HashMap::new();
        };
        let Some(plain) = decrypt(&content) else {
            return HashMap::new();
        };
        serde_json::from_str(&plain).unwrap_or_default()
    }

    fn load(&self, key: &str) -> Option<DiscoveryRecord> {
        self.read_all().remove(key)
    }

    fn store(&self, key: &str, record: &DiscoveryRecord) {
        let mut all = self.read_all();
        all.insert(key.to_string(), record.clone());
        let Ok(json) = serde_json::to_string_pretty(&all) else {
            return;
        };
        if let Some(parent) = self.path.parent() {
            let _ = std::fs::create_dir_all(parent);
        }
        let _ = std::fs::write(&self.path, encrypt(&json));
    }
}

fn build_discovery_url(base: &str) -> String {
    let base = base.trim_end_matches('/');
    if base.ends_with("/api/discovery/tenant") {
        base.to_string()
    } else if base.ends_with("/api/discovery") {
        format!("{base}/tenant")
    } else if base.ends_with("/api") {
        format!("{base}/discovery/tenant")
    } else {
        format!("{base}/api/discovery/tenant")
    }
}

async fn fetch_remote(
    base_url: &str,
    id_token: &str,
    tenant_hint: Option<&str>,
) -> Result<DiscoveryRecord, DiscoveryError> {
    let url = build_discovery_url(base_url);
    let mut body = serde_json::Map::new();
    if let Some(hint) = tenant_hint {
        body.insert("tenant_hint".into(), serde_json::Value::String(hint.to_string()));
    }
    let resp = reqwest::Client::new()
        .post(&url)
        .header("Authorization", format!("Bearer {id_token}"))
        .header("Content-Type", "application/json")
        .header("User-Agent", format!("kumiho-rust/{}", env!("CARGO_PKG_VERSION")))
        .json(&body)
        .send()
        .await
        .map_err(|e| DiscoveryError(format!("request failed: {e}")))?;
    let status = resp.status();
    if status.as_u16() >= 400 {
        let text = resp.text().await.unwrap_or_default();
        return Err(DiscoveryError(format!(
            "discovery endpoint returned {}: {}",
            status,
            &text.chars().take(200).collect::<String>()
        )));
    }
    resp.json::<DiscoveryRecord>()
        .await
        .map_err(|e| DiscoveryError(format!("invalid discovery payload: {e}")))
}

/// Resolve a [`DiscoveryRecord`], using the encrypted cache when fresh.
pub async fn resolve(
    id_token: &str,
    tenant_hint: Option<&str>,
    force_refresh: bool,
) -> Result<DiscoveryRecord, DiscoveryError> {
    let base_url =
        std::env::var("KUMIHO_CONTROL_PLANE_URL").unwrap_or_else(|_| DEFAULT_CONTROL_PLANE.to_string());
    let cache = DiscoveryCache { path: cache_path() };
    let cache_key = tenant_hint.unwrap_or(DEFAULT_CACHE_KEY);

    if !force_refresh {
        if let Some(cached) = cache.load(cache_key) {
            if !cached.cache_control.is_expired() {
                if cached.cache_control.should_refresh() {
                    match fetch_remote(&base_url, id_token, tenant_hint).await {
                        Ok(fresh) => {
                            cache.store(cache_key, &fresh);
                            return Ok(fresh);
                        }
                        Err(_) if !cached.cache_control.is_expired() => return Ok(cached),
                        Err(e) => return Err(e),
                    }
                }
                return Ok(cached);
            }
        }
    }

    let fresh = fetch_remote(&base_url, id_token, tenant_hint).await?;
    cache.store(cache_key, &fresh);
    Ok(fresh)
}

/// Probe loopback ports for a self-hosted CE server; return a gRPC target.
pub async fn resolve_local_ce_endpoint() -> Option<String> {
    let candidates: Vec<String> = if let Ok(ep) = std::env::var("KUMIHO_LOCAL_SERVER_ENDPOINT") {
        if ep.trim().is_empty() {
            vec![]
        } else {
            vec![normalize_local_target(ep.trim())]
        }
    } else if let Ok(port) = std::env::var("KUMIHO_LOCAL_SERVER_PORT") {
        match port.trim().parse::<u16>() {
            Ok(p) => vec![format!("127.0.0.1:{p}")],
            Err(_) => vec![],
        }
    } else {
        vec![format!("127.0.0.1:{DEFAULT_LOCAL_CE_PORT}")]
    };

    for target in candidates {
        if probe_ce(&target).await {
            return Some(target);
        }
    }
    None
}

fn normalize_local_target(raw: &str) -> String {
    let raw = raw.trim();
    let stripped = raw.split("://").last().unwrap_or(raw);
    if stripped.contains(':') {
        stripped.to_string()
    } else {
        format!("{stripped}:{DEFAULT_LOCAL_CE_PORT}")
    }
}

async fn probe_ce(target: &str) -> bool {
    let url = format!("http://{target}/api/_live");
    let client = match reqwest::Client::builder()
        .timeout(std::time::Duration::from_millis(500))
        .build()
    {
        Ok(c) => c,
        Err(_) => return false,
    };
    let Ok(resp) = client.get(&url).send().await else {
        return false;
    };
    if resp.status().as_u16() >= 400 {
        return false;
    }
    match resp.json::<serde_json::Value>().await {
        Ok(body) => body.get("deployment_mode").and_then(|v| v.as_str()) == Some("self_hosted_ce"),
        Err(_) => false,
    }
}
